extensions/data-plane/data-plane-aws-s3/src/main/java/org/eclipse/edc/connector/dataplane/aws/s3/S3DataSource.java

```diff
@@ -29,15 +35,69 @@ class S3DataSource implements DataSource {
 
     private String bucketName;
     private String keyName;
+    private String keyPrefix;
     private S3Client client;
 
-    private S3DataSource() { }
+    private S3DataSource() {
+    }
 
     @Override
     public StreamResult<Stream<Part>> openPartStream() {
+
+        if (keyPrefix != null) {
+            try {
+
+                var s3PartStream = this.fetchPrefixedS3Objects().stream()
+                        .map(S3Object::key)
+                        .map(key -> (Part) new S3Part(client, key, bucketName));
+
+                return success(s3PartStream);
+
+            } catch (S3Exception e) {
+                throw new RuntimeException("Error listing objects in the bucket: " + e.getMessage());
+            }
+        }
+
         return success(Stream.of(new S3Part(client, keyName, bucketName)));
     }
 
+    /**
+     * Fetches S3 objects with the specified prefix.
+     *
+     * @return A list of S3 objects.
+     * @throws S3ObjectNotFoundException if no objects are found.
+     */
+    private List<S3Object> fetchPrefixedS3Objects() {
+
+        String continuationToken = null;
+        List<S3Object> s3Objects = new ArrayList<>();
+
+        do {
+
+            var listObjectsRequest = ListObjectsV2Request.builder()
+                    .bucket(bucketName)
+                    .prefix(keyPrefix)
+                    .continuationToken(continuationToken)
+                    .build();
+
+            var response = client.listObjectsV2(listObjectsRequest);
+
+            s3Objects.addAll(response.contents());
+
+            continuationToken = response.nextContinuationToken();
+
+        } while (continuationToken != null);
+
+        if (s3Objects.isEmpty()) throw new S3ObjectNotFoundException("Object not found");
```

Comment:
I don't see the point of throwing an exception here that's caught right in the caller, please move this check on the caller